### PR TITLE
[framework] Move coin register events to account

### PR DESCRIPTION
**Pull request by davidiw**
This change will result in reducing all resources created by default by 1/3!
Due to circular dependencies coin <-> account cannot know about each
other, so umm... coins...

various tests validate it still works!

also we aren't using this event in our own indexer ... so I guess that makes my life easier.

Okay, so the overview:
* No one can call coin::register or account::register_coin, except their friends
* account knows about coin, but coin doesn't know about account
* account can have a friend on coin
* so in theory coins can be merged into account

I decided against that, because this is step 1. In step 2, I will move all resource oriented coin functions into `coins` and then redirect AptosCoin to account and other coins to coin.

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2189)

### CODE REVIEW

**Comment by davidiw** [comment]
Spoke offline with @movekevin , he does point out, rightfully so everything could be placed into Account, but I feel with such special logic, it will reduce discoverability.

**Comment by davidiw** [comment]
Not surprising... looks like our txn emitter is in goodish state :)

**Comment by msmouse** [comment]
after ```2022-07-25T03:52:26.903807Z [txn_committer] INFO execution/executor-benchmark/src/transaction_committer.rs:114 Version: 10021002. latency: 2943 ms, execute time: 1055 ms. commit time: 1117 ms. TPS: 8953. Accumulative TPS: 7938
2022-07-25T03:52:26.903880Z [txn_committer] INFO execution/executor-benchmark/src/transaction_committer.rs:123 Accumulative total: VM time: 447 secs, executor time: 622 secs, commit time: 1255 secs, DB commit time: 1023 secs
2022-07-25T03:52:26.903918Z [txn_committer] INFO execution/executor-benchmark/src/transaction_committer.rs:131 Accumulative per transaction: VM time: 44624 ns, executor time: 62114 ns, commit time: 125202 ns, DB commit time: 102070 ns
2022-07-25T03:52:28.293475Z [state_committer] INFO execution/executor-benchmark/src/state_committer.rs:112 Committing state. Saved.
Final state commit...
2022-07-25T03:52:28.765802Z [state_committer] INFO execution/executor-benchmark/src/state_committer.rs:91 Committing state. {"num_pending_commits":3,"version":10021002}
2022-07-25T03:52:32.009293Z [state_committer] INFO execution/executor-benchmark/src/state_committer.rs:112 Committing state. Saved.
Created 10000000 new accounts. Now at version 10021002, total # of accounts 10000000.```

**Comment by msmouse** [comment]
DB size (without pruner): 
```
msmouse@linux-dev:~/work/data/per-block$ du */* -sh
14G     10Mdavid/ledger_db
33G     10Mdavid/state_merkle_db
14G     10Mmain/ledger_db
38G     10Mmain/state_merkle_db
```

**Comment by davidiw** [comment]
> DB size (without pruner):
> 
> ```
> msmouse@linux-dev:~/work/data/per-block$ du */* -sh
> 14G     10Mdavid/ledger_db
> 33G     10Mdavid/state_merkle_db
> 14G     10Mmain/ledger_db
> 38G     10Mmain/state_merkle_db
> ```

storage isn't cheap! but still a nice 13% savings...

File: aptos-move/framework/aptos-framework/sources/coins.move [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    /// Script function to register to receive a specific `CoinType`. An account that wants to hold a coin type
+    /// has to explicitly registers to do so. The register creates a special `CoinStore`
+    /// to hold the specified `CoinType`.
+    public entry fun register<CoinType>(account: &signer) {
+        register_internal<CoinType>(account);
+    }
+
+    public fun register_internal<CoinType>(account: &signer) {
+        coin::register<CoinType>(account);
+				account::register_coin<CoinType>(signer::address_of(account));
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/coins.move:17]: god damn tabs

File: aptos-move/framework/aptos-framework/sources/coins.move [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/// This module allows for more convenient managing of coins across coin::CoinStore and
+/// account::Account
+module aptos_framework::coins {
+		use std::signer;
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/coins.move:4]: more tabbing

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -393,6 +403,20 @@ module aptos_framework::account {
...
+    public(friend) fun register_coin<CoinType>(account_addr: address) acquires Account {
+        let account = borrow_global_mut<Account>(account_addr);
+        event::emit_event<CoinRegisterEvent>(
+            &mut account.coin_register_events,
+            CoinRegisterEvent {
+                type_info: type_info::type_of<CoinType>(),
+            },
+        );
+    }
+
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/account.move:419]: Interesting, are more things going to be on the account?

**Review by gregnazario — APPROVED** [review]
Looks good to me, just fix the tabbing

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:419]: yes, the intention is to see if we can't get AptosCoin there as well... that will return p2p transactions to be the same cost as before fine grained storage.

File: aptos-move/e2e-testsuite/src/tests/data_store.rs [hunk 4]
```diff
@@ -111,15 +112,18 @@ fn borrow_after_move() {
         Ok(ExecutionStatus::ExecutionFailure { .. })
     ));
     executor.apply_write_set(output.write_set());
+    println!("HERE!");
```
Comment (movekevin) [inline aptos-move/e2e-testsuite/src/tests/data_store.rs:115]: Reminder to remove these println!'s.

File: aptos-move/framework/aptos-framework/sources/coins.move [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/// This module allows for more convenient managing of coins across coin::CoinStore and
+/// account::Account
+module aptos_framework::coins {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/coins.move:3]: Can we name this wallet instead of coins?

**Review by movekevin — APPROVED** [review]

File: aptos-move/e2e-testsuite/src/tests/data_store.rs (hunk 4 above)
Comment (davidiw) [reply aptos-move/e2e-testsuite/src/tests/data_store.rs:115]: well this is embarrassing ....

**Comment by davidiw** [comment]
I missed your comments somehow...

**Pull request merged**